javalib/src/main/scala/java/lang/Math.scala

```diff
@@ -1,324 +1,347 @@
+
+  @inline def incrementExact(a: scala.Long): scala.Long =
+    addExact(a, 1L)
+
+  @inline def log(a: scala.Double): scala.Double =
+    `llvm.log.f64`(a)
+
+  @inline def log10(a: scala.Double): scala.Double =
+    `llvm.log10.f64`(a)
+
+  @inline def log1p(a: scala.Double): scala.Double =
+    log(a + 1)
+
+  @inline def max(a: scala.Double, b: scala.Double): scala.Double =
+    `llvm.maxnum.f64`(a, b)
+
+  @inline def max(a: scala.Float, b: scala.Float): scala.Float =
+    `llvm.maxnum.f32`(a, b)
+
+  @inline def max(a: scala.Int, b: scala.Int): scala.Int =
+    if (a > b) a else b
+
+  @inline def max(a: scala.Long, b: scala.Long): scala.Long =
+    if (a > b) a else b
+
+  @inline def min(a: scala.Double, b: scala.Double): scala.Double =
+    `llvm.maxnum.f64`(a, b)
+
+  @inline def min(a: scala.Float, b: scala.Float): scala.Float =
+    `llvm.minnum.f32`(a, b)
+
+  @inline def min(a: scala.Int, b: scala.Int): scala.Int =
+    if (a < b) a else b
+
+  @inline def min(a: scala.Long, b: scala.Long): scala.Long =
+    if (a < b) a else b
+
+  @inline def multiplyExact(a: scala.Int, b: scala.Int): scala.Int = {
+    val overflow = `llvm.smul.with.overflow.i32`(a, b)
+    if (overflow.flag) throw new ArithmeticException("Integer overflow")
+    else overflow.value
   }
 
-  def tanh(a: scala.Double): scala.Double = {
-    if (a.isNaN || a == 0.0)
-      a
-    else if (abs(a) == scala.Double.PositiveInfinity)
-      signum(a)
-    else {
-      // sinh(a) / cosh(a) =
-      // 1 - 2 * (exp(-a)/ (exp(-a) + exp (a)))
-      val expma = exp(-a)
-      if (expma == scala.Double.PositiveInfinity) //Infinity / Infinity
-        -1.0
-      else {
-        val expa = exp(a)
-        val ret  = expma / (expa + expma)
-        1.0 - (2.0 * ret)
-      }
-    }
+  @inline def multiplyExact(a: scala.Long, b: scala.Long): scala.Long = {
+    val overflow = `llvm.smul.with.overflow.i64`(a, b)
+    if (overflow.flag) throw new ArithmeticException("Long overflow")
+    else overflow.value
   }
 
-  def addExact(a: scala.Int, b: scala.Int): scala.Int = {
-    val res       = a + b
-    val resSgnBit = res < 0
-    if (resSgnBit == (a < 0) || resSgnBit == (b < 0)) res
-    else throw new ArithmeticException("Integer overflow")
+  @inline def negateExact(a: scala.Int): scala.Int =
+    subtractExact(0, a)
+
+  @inline def negateExact(a: scala.Long): scala.Long =
+    subtractExact(0, a)
+
+  def nextAfter(a: scala.Float, b: scala.Double): scala.Float = {
+    if (Float.isNaN(a) || Double.isNaN(b)) {
+      Float.NaN
+    } else if (abs(a) == 0f && abs(b) == 0d) {
+      b.toFloat
+    } else if (abs(a) == Float.MIN_VALUE && abs(b) < abs(a.toDouble)) {
+      copySign(0, a)
+    } else if (Float.isInfinite(a) && abs(b) < abs(a.toDouble)) {
+      copySign(Float.MAX_VALUE, a)
+    } else if (abs(a) == Float.MAX_VALUE && abs(b) > abs(a.toDouble)) {
+      copySign(Float.POSITIVE_INFINITY, a)
+    } else {
+      cmath.nextafter(a.toDouble, b).toFloat
```

Comment:
I am not sure this is correct. In fact, I'm pretty sure it isn't.

Since doubles have a much higher precision, `nextafter(a.toDouble, b)` will be 1 up *in terms of doubles*. Converting that back into float will probably just round it down to the original float.